File: python3/leetcodepy/candy.py
"""
There are N children standing in a line. Each child is assigned a rating value.

You are giving candies to these children subjected to the following requirements:

Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.

What is the minimum candies you must give?

Example 1:

Input: [1,0,2]
Output: 5
Explanation: You can allocate to the first, second and third child with 2, 1, 2 candies respectively.

Example 2:

Input: [1,2,2]
Output: 4
Explanation: You can allocate to the first, second and third child with 1, 2, 1 candies respectively.
             The third child gets 1 candy because it satisfies the above two conditions.
"""
from typing import List


class Solution1:
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        size = len(ratings)
        candies = [1] * size
        for i in range(1, size):
            if ratings[i] > ratings[i - 1]:
                candies[i] = candies[i - 1] + 1
        for i in range(size - 2, -1, -1):
            if ratings[i] > ratings[i + 1]:
                candies[i] = max(candies[i], candies[i + 1] + 1)
        return sum(candies)

File: java/src/main/java/io/binac/leetcode/Candy.java
package io.binac.leetcode;

import java.util.Arrays;

/**
 * <p>There are <code>n</code> children standing in a line. Each child is assigned a rating value given in the integer array <code>ratings</code>.</p>
 *
 * <p>You are giving candies to these children subjected to the following requirements:</p>
 *
 * <ul>
 * 	<li>Each child must have at least one candy.</li>
 * 	<li>Children with a higher rating get more candies than their neighbors.</li>
 * </ul>
 *
 * <p>Return <em>the minimum number of candies you need to have to distribute the candies to the children</em>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> ratings = [1,0,2]
 * <strong>Output:</strong> 5
 * <strong>Explanation:</strong> You can allocate to the first, second and third child with 2, 1, 2 candies respectively.
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> ratings = [1,2,2]
 * <strong>Output:</strong> 4
 * <strong>Explanation:</strong> You can allocate to the first, second and third child with 1, 2, 1 candies respectively.
 * The third child gets 1 candy because it satisfies the above two conditions.
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>n == ratings.length</code></li>
 * 	<li><code>1 &lt;= n &lt;= 2 * 10<sup>4</sup></code></li>
 * 	<li><code>0 &lt;= ratings[i] &lt;= 2 * 10<sup>4</sup></code></li>
 * </ul>
 */
public class Candy {
    public static class Solution1 {
        public int candy(int[] ratings) {
            int[] candies = new int[ratings.length];
            Arrays.fill(candies, 1);
            for (int i = 1; i < ratings.length; ++i) {
                if (ratings[i] > ratings[i - 1])
                    candies[i] = candies[i - 1] + 1;
            }
            int result = candies[ratings.length - 1];
            for (int i = ratings.length - 2; i >= 0; --i) {
                if (ratings[i] > ratings[i + 1])
                    candies[i] = Math.max(candies[i], candies[i + 1] + 1);
                result += candies[i];
            }
            return result;
        }
    }
}
